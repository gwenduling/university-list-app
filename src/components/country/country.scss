.country-component {
  margin-top: 20px;
}

.result-count {
  font-size: 14px;
}

.show-more {
  width: calc(25% - 10px);

  @media (max-width: 1300px) {
    width: calc(33% - 3px);
  }

  @media (max-width: 500px) {
    width: 100%;
  }
}
