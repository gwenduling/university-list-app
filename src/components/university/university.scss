.university-component {
  padding: 10px;
  color: var(--color-darker);
  background-color: var(--color-light);

  .name {
    margin: 0;
    font-weight: 700;
  }

  .link {
    color: blue;
    font-size: 12px;
  }
}
