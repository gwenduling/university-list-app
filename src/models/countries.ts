export interface CountryModel {
  alpha2Code: string;
  altSpellings: string[];
  flag: string;
  name: string;
  nativeName: string;
  count: number;
  total: number;
}

export interface CountryResponseModel {
  alpha2Code: string;
  alpha3Code: string;
  altSpellings: string[];
  capital: string;
  demonym: string;
  flag: string;
  name: string;
  nativeName: string;
}
