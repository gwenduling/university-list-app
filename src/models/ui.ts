export enum StateUI {
  Loading = "loading",
  Success = "success",
  Error = "error",
}
