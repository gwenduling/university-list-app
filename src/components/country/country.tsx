import { Component } from "react";
import { CountryModel } from "../../models/countries";
import { UniversityModel } from "../../models/universities";

import University from "../university/university";
import CountryItem from "../country-item/country-item";

import "./country.scss";

class Country extends Component<Props> {
  renderUniversity(university: UniversityModel) {
    return (
      <University
        university={university}
        key={university.alpha_two_code + "" + university.name}
      />
    );
  }

  renderButton() {
    return (
      <button
        className="show-more"
        onClick={() => this.props.onClick?.(this.props.country.alpha2Code)}
      >
        show more
      </button>
    );
  }

  render() {
    const universities = this.props.universities
      .filter(
        (university: UniversityModel) =>
          university.alpha_two_code === this.props.country.alpha2Code
      )
      .slice(0, this.props.country.count);

    let button;
    if (this.props.country.count < this.props.country.total) {
      button = this.renderButton();
    } else {
      button = "";
    }

    return (
      <div className="country-component">
        <CountryItem country={this.props.country} />

        <div className="result-count">
          Showing {this.props.country.count} of {this.props.country.total}{" "}
          universities
        </div>

        <div className="universities-list">
          {universities.map((university: UniversityModel) => {
            return this.renderUniversity(university);
          })}
        </div>

        {button}
      </div>
    );
  }
}

interface Props {
  country: CountryModel;
  universities: UniversityModel[];
  onClick?: (code: string) => void;
}
export default Country;
