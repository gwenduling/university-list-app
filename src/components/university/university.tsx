import { UniversityModel } from "../../models/universities";

import "./university.scss";

function University(props: Props) {
  return (
    <div className="university-component">
      <p className="name">{props.university.name}</p>
      <a
        className="link"
        href={props.university.web_pages[0]}
        target="_blank"
        rel="noreferrer"
        title={props.university.name}
      >
        {props.university.domains[0]}
      </a>
    </div>
  );
}

interface Props {
  university: UniversityModel;
}
export default University;
