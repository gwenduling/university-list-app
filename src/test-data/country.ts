import { CountryModel } from "../models/countries";
export const mockCountry: CountryModel = {
  alpha2Code: "PH",
  altSpellings: ["PH", "Republic of the Philippines", "Rep√∫blika ng Pilipinas"],
  flag: "https://restcountries.eu/data/phl.svg",
  name: "Philippines",
  nativeName: "Pilipinas",
  count: 0,
  total: 0,
};

export const mockCountry2: CountryModel = {
  alpha2Code: "US",
  altSpellings: ["US", "USA", "United States of America"],
  flag: "https://restcountries.eu/data/phl.svg",
  name: "United States of America",
  nativeName: "United States",
  count: 0,
  total: 0,
};
