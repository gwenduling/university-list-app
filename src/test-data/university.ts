import { UniversityModel } from "../models/universities";

export const mockUniversity: UniversityModel = {
  alpha_two_code: "PH",
  country: "Philippines",
  name: "University Test 1",
  domains: ["university.com"],
  web_pages: ["http://university.com"],
  "state-province": null,
};

export const mockUniversity2: UniversityModel = {
  alpha_two_code: "PH",
  country: "Philippines",
  name: "University Test 2",
  domains: ["university2.com"],
  web_pages: ["http://university2.com"],
  "state-province": null,
};
