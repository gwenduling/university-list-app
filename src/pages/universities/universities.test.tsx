import Universities from "./universities";
import Country from "../../components/country/country";
import Search from "../../components/search/search";
import Loading from "../../components/loading/loading";
import Error from "../../components/error/error";
import flushPromises from "flush-promises";

import { shallow, mount } from "enzyme";
import { StateUI } from "../../models/ui";
import { mockUniversity, mockUniversity2 } from "../../test-data/university";
import { mockCountry, mockCountry2 } from "../../test-data/country";

global.fetch = jest.fn(() => Promise.reject());
console.error = jest.fn();

describe("Universities Component", () => {
  let wrapper: any;
  beforeEach(() => {
    wrapper = shallow(<Universities />);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render error UI", () => {
    wrapper.setState({ stateUI: StateUI.Error });

    expect(wrapper.find(Error)).toHaveLength(1);
    expect(wrapper.find(Loading)).toHaveLength(0);
  });

  it("should trigger error callback", () => {
    const mountWrapper = mount(<Universities />);

    // @ts-ignore
    mountWrapper.instance().fetchUniversities = jest.fn();
    mountWrapper.setState({ stateUI: StateUI.Error });
    mountWrapper.find("button").simulate("click");
    // @ts-ignore
    expect(mountWrapper.instance().fetchUniversities).toHaveBeenCalledTimes(1);
  });

  it("should render loading UI", () => {
    wrapper.setState({ stateUI: StateUI.Loading });

    expect(wrapper.find(Error)).toHaveLength(0);
    expect(wrapper.find(Loading)).toHaveLength(1);
  });

  it("should render search", () => {
    wrapper.setState({ stateUI: StateUI.Success });

    expect(wrapper.find(Error)).toHaveLength(0);
    expect(wrapper.find(Loading)).toHaveLength(0);
    expect(wrapper.find(Search)).toHaveLength(2);
  });

  it("should render countries", () => {
    wrapper.setState({
      stateUI: StateUI.Success,
      countries: [
        { ...mockCountry, count: 1, total: 1 },
        { ...mockCountry2, count: 1, total: 1 },
      ],
      universities: [mockUniversity, mockUniversity2],
    });

    expect(wrapper.find(Country)).toHaveLength(2);
  });

  it("should trigger show more", () => {
    const mountWrapper = mount(<Universities />);

    // @ts-ignore
    mountWrapper.instance().showMore = jest.fn();
    mountWrapper.setState({
      stateUI: StateUI.Success,
      countries: [{ ...mockCountry, count: 9, total: 20 }],
      universities: [],
    });
    mountWrapper.find("button").simulate("click");
    // @ts-ignore
    expect(mountWrapper.instance().showMore).toHaveBeenCalledTimes(1);
  });

  it("should add 9 universities in UI on show more", () => {
    wrapper.setState({
      stateUI: StateUI.Success,
      countries: [{ ...mockCountry, count: 9, total: 20 }],
      universities: [],
    });

    wrapper.instance().showMore("PH");
    expect(wrapper.state("countries")[0].count).toBe(18);
  });

  it("should set count to total count on show more if remaining is less than 9", () => {
    wrapper.setState({
      stateUI: StateUI.Success,
      countries: [{ ...mockCountry, count: 9, total: 12 }],
      universities: [],
    });

    wrapper.instance().showMore("PH");
    expect(wrapper.state("countries")[0].count).toBe(12);
  });

  it("should filter countries", () => {
    wrapper.setState({
      stateUI: StateUI.Success,
      countries: [
        { ...mockCountry2, count: 1, total: 1 },
        { ...mockCountry, count: 1, total: 1 },
      ],
      universities: [mockUniversity, mockUniversity2],
      searchCountry: "PH",
    });

    expect(wrapper.find(Country)).toHaveLength(1);
  });

  it("should filter out countries with 0 total", () => {
    wrapper.setState({
      stateUI: StateUI.Success,
      countries: [mockCountry2],
      universities: [],
    });

    expect(wrapper.find(Country)).toHaveLength(0);
  });

  it("should update country search value", () => {
    wrapper.setState({
      stateUI: StateUI.Success,
      country: mockCountry,
      universityList: [],
    });
    wrapper.instance().handleCountrySearchChange("test");
    expect(wrapper.state("searchCountry")).toBe("test");
  });

  it("should update university search value", () => {
    jest.spyOn(wrapper.instance(), "searchUniversities");
    wrapper.setState({
      stateUI: StateUI.Success,
      country: mockCountry,
      universityList: [],
    });
    wrapper.instance().handleUniversitySearchChange("test");
    expect(wrapper.state("searchUniversity")).toBe("test");
  });

  it("should filter universities", async () => {
    // @ts-ignore
    global.fetch = jest.fn(async () =>
      Promise.resolve({
        ok: true,
        json: async () => [mockUniversity],
      })
    );

    wrapper.setState({
      stateUI: StateUI.Success,
      countries: [
        { ...mockCountry2, count: 1, total: 1 },
        { ...mockCountry, count: 1, total: 1 },
      ],
      universities: [mockUniversity, mockUniversity2],
      searchUniversity: "PH",
    });

    expect(wrapper.state("universities")).toHaveLength(2);

    wrapper.instance().searchUniversities();
    await flushPromises();

    expect(wrapper.state("universities")).toHaveLength(1);
  });

  it("should fetch countries after universities fetch", async () => {
    jest.spyOn(wrapper.instance(), "fetchCountries");

    // @ts-ignore
    global.fetch = jest.fn(async () =>
      Promise.resolve({
        ok: true,
        json: async () => [mockUniversity, mockUniversity2],
      })
    );

    wrapper.instance().fetchUniversities();
    await flushPromises();

    expect(wrapper.instance().fetchCountries).toHaveBeenCalled();
  });

  it("should display success view after successful fetch", async () => {
    // @ts-ignore
    global.fetch = jest.fn(async () =>
      Promise.resolve({
        ok: true,
        json: async () => [mockCountry, mockCountry2],
      })
    );

    wrapper.instance().fetchCountries();
    await flushPromises();

    expect(wrapper.state("stateUI")).toBe(StateUI.Success);
  });

  it("should display error view after error fetch", async () => {
    // @ts-ignore
    global.fetch = jest.fn(async () => Promise.reject());

    wrapper.instance().fetchCountries();
    await flushPromises();

    expect(wrapper.state("stateUI")).toBe(StateUI.Error);
  });
});
