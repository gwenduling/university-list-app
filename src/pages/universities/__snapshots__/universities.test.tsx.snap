// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Universities Component should match the snapshot 1`] = `
<div
  className="universities _padding -grow"
>
  <ReactTitle
    title="Universities | University List App"
  />
  <Error
    content="universities"
    onClick={[Function]}
  />
</div>
`;
